tests/heka/test_message_parser.py

```diff
@@ -1,87 +1,121 @@
-                    count += 1
-            except Exception:
-                threw = True
-        assert expected_exceptions[t] == threw
-
-
-top_keys = set(["application", "clientId", "creationDate", "environment", "id", "meta",
-                "payload", "type", "version"])
-payload_keys = set(["UIMeasurements", "addonDetails", "childPayloads", "chromeHangs",
-                    "fileIOReports", "histograms", "info", "keyedHistograms", "lateWrites",
-                    "log", "processes", "simpleMeasurements", "slowSQL", "threadHangStats",
-                    "ver", "webrtc"])
-
-
-def test_telemetry(data_dir):
-    filename = "{}/test_telemetry_gzip.heka".format(data_dir)
+@pytest.mark.parametrize("heka_format,try_snappy,strict,expected_count,expected_exception", [
+    # snappy disabled
+    ("plain", False, False, 10, False),
+    ("snappy", False, False, 0, False),
+    ("mixed", False, False, 5, False),
+    ("gzip",  False, False, 10, False),
+    ("gzip_mixed", False, False, 5, False),
+    # snappy enabled
+    ("plain", True, False, 10, False),
+    ("snappy", True, False, 10, False),
+    ("mixed", True, False, 10, False),
+    ("gzip",  True, False, 10, False),
+    ("gzip_mixed", True, False, 10, False),
+    # strict mode on
+    ("plain", False, True, 10, False),
+    ("snappy", False, True, 0, True),
+    ("mixed", False, True, 5, True),
+    ("gzip",  False, True, 10, False),
+    ("gzip_mixed", False, True, 5, True)
+])
+def test_unpack(data_dir, heka_format, try_snappy, strict, expected_count,
+                expected_exception):
+    count = 0
+    threw_exception = False
+    filename = "{}/test_{}.heka".format(data_dir, heka_format)
     with open(filename, "rb") as o:
-        for r in message_parser.parse_heka_message(streaming_gzip_wrapper(o)):
-            assert set(r.keys()) == top_keys
-            assert set(r["payload"].keys()) == payload_keys
+        if "gzip" in heka_format:
+            o = streaming_gzip_wrapper(o)
+        try:
+            for r, b in message_parser.unpack(o, try_snappy=try_snappy, strict=strict):
+                j = json.loads(r.message.payload)
+                assert count == j["seq"]
+                count += 1
+        except DecodeError:
+            threw_exception = True
 
-    filename = "{}/test_telemetry_snappy.heka".format(data_dir)
-    with open(filename, "rb") as o:
-        for r in message_parser.parse_heka_message(o):
-            assert set(r.keys()) == top_keys
-            assert set(r["payload"].keys()) == payload_keys
+    assert count == expected_count
+    assert threw_exception == expected_exception
+
+
+@pytest.mark.parametrize("heka_format", ["snappy", "gzip"])
+def test_telemetry(data_dir, heka_format):
+    filename = "{}/test_telemetry_{}.heka".format(data_dir, heka_format)
+    reference_filename = filename + '.json'
+
+    # enable this to regenerate the expected json representation of the ping
+    if False:
```

Comment:
I feel like we want generating the file to be a conscious choice, so I think I'm going to leave this as-is for now.